src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy

```diff
@@ -0,0 +1,115 @@
+package org.dormitory.autobotsoccub.engine
+
+import org.dormitory.autobotsoccub.engine.exception.GameDoesNotExistException
+import org.dormitory.autobotsoccub.engine.model.Game
+import org.dormitory.autobotsoccub.engine.model.GameData
+import spock.lang.Specification
+
+import static org.dormitory.autobotsoccub.engine.model.PlayPosition.GK
+
+class GameEngineShould extends Specification {
+
+    def GK_A = 0
+    def FW_A = 1
+    def GK_B = 2
+    def FW_B = 3
+
+    Game game
+    GameEngine sut = new GameEngine()
+
+    def setup() {
+        game = Game.builder()
+                .goalKeeperA(GK_A)
+                .forwardA(FW_A)
+                .goalKeeperB(GK_B)
+                .forwardB(FW_B)
+                .build()
+    }
+
+    def "correctly started game"() {
+        when:
+        def newGame = sut.startGame(FW_A, game)
+        def existGame = sut.startGame(FW_A, game)
+
+        then:
+        newGame
+        !existGame
+    }
+
+    def "correctly executed game commands"() {
+        given:
+        sut.startGame(FW_A, game)
+        sut.metaClass.getCurrentGame << { Integer userId ->
+            GameData activeGame = activeGames.get userId
+            GameData.builder()
+                    .game(activeGame.getGame())
+                    .scoreTable(activeGame.getScoreTable())
+                    .build()
+        }
+
+        //correctly count score && count auto score && changed position in the team
+        when:
+        sut.score FW_A
+        sut.autoScore FW_A
+        sut.changePositionsInTeam FW_A
+        def forwardAStats = sut.getCurrentGame(FW_A).scoreTable.statsByUserId.get(FW_A)
+
+        then:
+        forwardAStats.scored == 1
+        forwardAStats.autoScored == 1
+        forwardAStats.position == GK
+
+        //correctly catch exception
+        when:
+        sut.score 999
+
+        then:
+        GameDoesNotExistException e = thrown()
+        e.cause == null
+
+        //correctly reverted last event & stop game
+        when:
```

Comment:
1. "correctly count scored goal" (use dataTable to supply different scores (per user) that should be verified) 
2. "fail to score" (think about edge cases)

add corresponding tests for autoScore and revert

Reply:
In this section of the tests, the "Revert" command is checked, the extreme case is leaned - cancel the last event when it was already canceled. (This is the only extreme case for this test case)